lib/spack/spack/util/timer.py

```diff
@@ -0,0 +1,54 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+"""Debug signal handler: prints a stack trace and enters interpreter.
+
+``register_interrupt_handler()`` enables a ctrl-C handler that prints
+a stack trace and drops the user into an interpreter.
+
+"""
+import spack.util.spack_json as sjson
+import time
+import sys
+
+
+class Timer(object):
+    """
+    Simple timer for timing phases of a solve or install
+    """
+    def __init__(self):
+        self.start = time.time()
+        self.last = self.start
+        self.phases = {}
+        self.total = None
+        self.end = None
+
+    def phase(self, name):
+        last = self.last
+        now = time.time()
+        self.phases[name] = now - last
+        self.last = now
+
+    def stop(self):
+        """
+        Stop the timer to record a total time, if desired.
+        """
+        self.end = time.time()
+        self.total = self.end - self.start
```

Comment:
avoid duplicating the logic for `total` and just make it a phase with a special name like `_spack_timer_total` (in case some client code wants `"total"`)

Reply:
Will do! But what do you mean by some client code wanting total?

Reply:
I just meant that if someone uses this ("client code" is code that uses this) they might want a phase to be called `"total"`, so we should use some name for it that nobody's really likely to want to use.